File: assessment/challenge-03-practical/src/services/checkInService.ts
// checkInService.ts - Check-In Service with Duplicate Prevention
import { CheckInResult } from '../models/types';
import { globalRepository, EventHubRepository } from '../lib/db';

export class CheckInService {
  constructor(private repo: EventHubRepository = globalRepository) {}

  public parseTicketId(rawInput: any): string {
    if (rawInput === undefined || rawInput === null) return '';
    const clean = String(rawInput).trim();
    if (clean.startsWith('GDG-PASS:')) {
      return clean.replace('GDG-PASS:', '').trim();
    }
    return clean;
  }

  public checkIn(
    rawInput: any,
    scannedBy: string = 'ORGANIZER_DESK',
    deviceInfo?: string
  ): CheckInResult {
    const ticketId = this.parseTicketId(rawInput);
    if (!ticketId) {
      return {
        success: false,
        status: 'INVALID_TICKET',
        message: 'Missing or empty ticket ID',
      };
    }

    const res = this.repo.performCheckIn(ticketId, scannedBy, deviceInfo);

    if (res.status === 'VALID_TICKET' && res.attendee) {
      return {
        success: true,
        status: 'VALID_TICKET',
        message: `Welcome ${res.attendee.fullName}! Check-in verified.`,
        ticketId: res.ticket ? res.ticket.id : ticketId,
        attendee: {
          fullName: res.attendee.fullName,
          email: res.attendee.email,
          rollNumber: res.attendee.rollNumber,
          department: res.attendee.department,
          year: res.attendee.year,
        },
        checkInTime: res.scannedAt,
      };
    } else if (res.status === 'ALREADY_CHECKED_IN') {
      return {
        success: false,
        status: 'ALREADY_CHECKED_IN',
        message: 'Security Alert: Ticket has ALREADY BEEN CHECKED IN',
        ticketId: res.ticket ? res.ticket.id : ticketId,
        attendee: res.attendee
          ? {
              fullName: res.attendee.fullName,
              email: res.attendee.email,
              rollNumber: res.attendee.rollNumber,
              department: res.attendee.department,
              year: res.attendee.year,
            }
          : undefined,
      };
    } else {
      return {
        success: false,
        status: 'INVALID_TICKET',
        message: `Ticket '${ticketId}' was not found in database`,
        ticketId,
      };
    }
  }
}

File: assessment/challenge-03-practical/src/services/checkin_service.py
# checkin_service.py - QR & Ticket Check-In Service with Duplicate Protection

import uuid
from typing import Dict, Any, Optional
from lib.database import EventHubDatabase

class CheckInService:
    def __init__(self, db: EventHubDatabase):
        self.db = db

    def extract_ticket_id(self, raw_input: str) -> str:
        """Parses ticket ID from raw string or QR pass URI format 'GDG-PASS:TICK-GDG-XXXX'."""
        if not raw_input:
            return ""
        clean = raw_input.strip()
        if clean.startswith("GDG-PASS:"):
            return clean.replace("GDG-PASS:", "").strip()
        return clean

    def check_in(
        self,
        raw_ticket_or_qr: str,
        scanned_by: str = "ORGANIZER_DESK",
        device_info: Optional[str] = None
    ) -> Dict[str, Any]:
        ticket_id = self.extract_ticket_id(raw_ticket_or_qr)
        if not ticket_id:
            return {
                "success": False,
                "status": "INVALID_TICKET",
                "message": "Ticket identifier is missing or malformed"
            }

        checkin_id = str(uuid.uuid4())
        status, row_data, err_msg = self.db.check_in_ticket(
            checkin_id=checkin_id,
            ticket_id=ticket_id,
            scanned_by=scanned_by,
            device_info=device_info
        )

        if status == "VALID_TICKET":
            return {
                "success": True,
                "status": "VALID_TICKET",
                "message": "Check-in successful! Welcome to GDG DevFest SVEC 4.0",
                "ticketId": ticket_id,
                "attendee": {
                    "fullName": row_data["full_name"],
                    "email": row_data["email"],
                    "rollNumber": row_data["roll_number"],
                    "department": row_data["department"],
                    "year": row_data["year"],
                    "status": "CHECKED_IN",
                    "scannedAt": row_data.get("scanned_at")
                }
            }
        elif status == "ALREADY_CHECKED_IN":
            return {
                "success": False,
                "status": "ALREADY_CHECKED_IN",
                "message": "Security Alert: This ticket has ALREADY BEEN CHECKED IN.",
                "ticketId": ticket_id,
                "attendee": {
                    "fullName": row_data["full_name"] if row_data else "Unknown",
                    "department": row_data["department"] if row_data else "Unknown",
                    "status": "ALREADY_CHECKED_IN"
                } if row_data else None
            }
        else:
            return {
                "success": False,
                "status": "INVALID_TICKET",
                "message": err_msg or "Ticket not found in registration database",
                "ticketId": ticket_id
            }

    def get_ticket_status(self, ticket_id: str) -> Optional[Dict[str, Any]]:
        clean_id = self.extract_ticket_id(ticket_id)
        return self.db.get_ticket_with_attendee(clean_id)
